File: C/Learn/recursions.c
#include<stdio.h>

int fun_hello(int count){

    if(count<=0){
        return 0;
    }

    printf("Hello\n");
    fun_hello(count-1);
    
}

int fun_sum_natural_num(int num){
    int sum;
    if(num==1){
        return 1;
    } 
    sum= num + fun_sum_natural_num(num-1);
    return sum;
}

int fun_factorial(int num){

    int fact;
    if(num==0){
        return 1;
    }
    fact=num*fun_factorial(num-1);
    return fact;

}

int fun_fibonacci(int n){
    if(n==0){
        return 0;
    }
    if(n==1){
        return 1;
    }
    int fibNm1=fun_fibonacci(n-1);
    int fibNm2=fun_fibonacci(n-2);
    int fibN=fibNm1 + fibNm2;
    
    return fibN;
}

int main(){

    fun_hello(5);
    
    int sum=fun_sum_natural_num(10);
    printf("Sum is %d\n",sum);

    int fact=fun_factorial(7);
    printf("factorial is : %d\n",fact);

    int fibonacci=fun_fibonacci(5);
    printf("fibonacci is : %d\n",fibonacci);

    return 0;
}

File: PYthon/Learn/recursions.py
def fun_show(n):
    if(n==0):      #Base Case
        return
    print(n)
    fun_show(n-1)
    print("END")

fun_show(9)

def fun_factorial(n):
    if(n==1 or n==0):
        return 1
    else:
        return n*fun_factorial(n-1)

print("factorial is : ",fun_factorial(4))

def fun_natural_num_sum(n):
    if(n==0):
        return 0
    else:
        return n+fun_natural_num_sum(n-1)

print("Sum is :",fun_natural_num_sum(10))


def fun_print_lists(lists,idx=0):
    
    if(idx==len(lists)):
        return
    else:
        print(lists[idx])
        idx+=1
        fun_print_lists(lists,idx)



lists= [1,2,3,"nil"]
fun_print_lists(lists)
